### PR TITLE
bugfix: fix pouch not restore containerd when containerd killed unexpectly

### PR DESCRIPTION
**Pull request by Ace-Tang**
fix pouch not restore containerd when it get accident，refactor
the way in wait task exit, detail is in
https://github.com/alibaba/pouch/issues/1460

Fixes: #1460

Signed-off-by: Ace-Tang <aceapril@126.com>

### CODE REVIEW

**Comment by Ace-Tang** [comment]
Detail design logic is locate at https://github.com/alibaba/pouch/issues/1460

**Comment by rudyfly** [comment]
ping @fuweid @HusterWan Can you review this pr?

File: ctrd/client.go [hunk 1]
```diff
@@ -328,3 +346,30 @@ func (c *Client) Cleanup() error {
 
 	return nil
 }
+
+// handleConnectContainerd restore containerd
+func (c *Client) handleConnectContainerd() {
+
+	for {
+		logrus.Info("handle restore containerd")
```
Comment (HusterWan) [inline ctrd/client.go]: > logrus.Info("handle restore containerd")

> logrus.Infof("ctrd: restore containerd")

should we only keep one log?

File: ctrd/client.go [hunk 2]
```diff
@@ -328,3 +346,30 @@ func (c *Client) Cleanup() error {
 
 	return nil
 }
+
+// handleConnectContainerd restore containerd
+func (c *Client) handleConnectContainerd() {
```
Comment (HusterWan) [inline ctrd/client.go]: If containerd process restart, we should also recreate new ctrd client,  WDYT?

**Comment by HusterWan** [comment]
@Ace-Tang Thanks a lot for your work to enhance the management of containerd process instance.

There is indeed a shortcut that when containerd process existed unexpected, we will not pull it up again before, so i think this pr is wonderful !!

there has one thing we should discuss, the old grpc client's stream client has 100 limitation, so we use a grpc clients pool to break the limitation.

so if we restart containerd, we also should notice the stream client counter.

File: ctrd/client.go (hunk 1 above)
Comment (Ace-Tang) [reply ctrd/client.go]: Sure.

File: ctrd/client.go (hunk 2 above)
Comment (Ace-Tang) [reply ctrd/client.go]: Maybe client pool need more optimization.

File: test/cli_exec_test.go [hunk 3]
```diff
@@ -143,3 +147,24 @@ func (suite *PouchExecSuite) TestExecExitCode(c *check.C) {
 	command.PouchRun("exec", name, "sh", "-c", "exit 101").Assert(c, icmd.Expected{ExitCode: 101})
 	command.PouchRun("exec", name, "sh", "-c", "exit 0").Assert(c, icmd.Success)
 }
+
+// TestExecWithContainerdKilled test containerd get unexpected killed, and will restore
+func (suite *PouchExecSuite) TestExecWithContainerdKilled(c *check.C) {
+	name := "TestExecWithContainerdKilled"
+	res := command.PouchRun("run", "-d", "--name", name, busyboxImage, "top")
+	res.Assert(c, icmd.Success)
```
Comment (Letty5411) [inline test/cli_exec_test.go]: add a `defer DelContainerForceMultyTime(c, name)`

File: test/cli_exec_test.go [hunk 4]
```diff
@@ -143,3 +147,24 @@ func (suite *PouchExecSuite) TestExecExitCode(c *check.C) {
...
+	// give time for containerd restore, more than 3 second.
+	data, err := exec.Command("pidof", "containerd").CombinedOutput()
+	c.Assert(err, check.IsNil)
+	pid, err := strconv.Atoi(strings.TrimSpace(string(data)))
+	c.Assert(err, check.IsNil)
+
+	syscall.Kill(pid, syscall.SIGTERM)
+	time.Sleep(5)
+
+	output := command.PouchRun("exec", name, "echo", "1").Stdout()
```
Comment (Letty5411) [inline test/cli_exec_test.go]: I would prefer to use:
```
res := command.PouchRun("exec", name, "echo", "1")
res.Assert(c, icmd.Success)
output := res.Stdout()
```

**Comment by Ace-Tang** [comment]
@Letty5411 , from the test result, seems I kill containerd in test will influence cri test, maybe another containerd is need for the test.

File: ctrd/client.go (hunk 2 above)
Comment (Ace-Tang) [reply ctrd/client.go]:
```
 gopts := []grpc.DialOption{
        grpc.WithBlock(),
        grpc.WithInsecure(),
        grpc.WithTimeout(60 * time.Second),
        grpc.FailOnNonTempDialError(true),
        grpc.WithBackoffMaxDelay(3 * time.Second),
        grpc.WithDialer(dialer.Dialer),
    }
```
from `containerd.New` grpc connect get a 60s timeout, assumes that we do not consider about containerd killed unexpectly, valid of these connect should be consider.

**Pull request closed without merging**